### ISSUE TITLE
Test for empty items in radios and checkboxes

### ISSUE THREAD

**Issue by edwardhorsford** [alphagov/govuk-frontend#1494]
I'd like to propose testing if items inside the `items` array for checkboxes and radios are empty. If they are, don't output markup.

## My use case
I want to conditionally show a radio based on some data.

In Nunjucks I can alter a value of an item in an array by adding a conditional after it - but it still outputs an empty item and I've not worked out how to not provide it.

I could prepare my data in advance - but for a simple thing like this, doing it directly is more readable IMHO. You can see the logic directly.

Example macro:
```js
{{ govukRadios({
  idPrefix: "where-do-you-live",
  name: "where-do-you-live",
  fieldset: {
    legend: {
      text: "Where do you live?",
      isPageHeading: true,
      classes: "govuk-fieldset__legend--xl"
    }
  },
  items: [
    {
      value: "england",
      text: "England"
    },
    {
      value: "scotland",
      text: "Scotland"
    },
    {
      value: "wales",
      text: "Wales"
    },
    {
      value: "northern-ireland",
      text: "Northern Ireland"
    },
    {
      divider: "or"
    } if data.showFantasy == true,
    {
      value: "narnia",
      text: "Narnia"
    } if data.showFantasy == true
  ]
}) }}
```

This _mostly_ works - but right now outputs a blank space and random html doing nothing:
<img width="631" alt="Screenshot 2019-07-12 at 15 23 36" src="https://user-images.githubusercontent.com/2204224/61135180-0c133080-a4b9-11e9-9d2a-3542121e440a.png">
<img width="486" alt="Screenshot 2019-07-12 at 15 24 12" src="https://user-images.githubusercontent.com/2204224/61135217-20572d80-a4b9-11e9-8faf-4f046799591e.png">

## Suggested fix
Either strip out empty items from the `items` array or else test whether they're empty when iterating over them.

This feels like a safer thing to do / more defensive. We don't output markup where there's no input.

I had a quick test of this by editing the radios macro directly in my node_modules, and it appears to work well.

**Comment by nickcolley** [comment]
I've put together a live demonstration to show what Ed is talking about here: https://glitch.com/edit/#!/govuk-frontend-issue-1494

**Comment by nickcolley** [comment]
Should we also consider doing something similar for:

- https://design-system.service.gov.uk/components/accordion/ (items)
- https://design-system.service.gov.uk/components/breadcrumbs/ (items)
- https://design-system.service.gov.uk/components/error-summary/ (error list)
- https://design-system.service.gov.uk/components/tabs/ (items)
- https://design-system.service.gov.uk/components/summary-list/ (rows)

**Comment by paulmsmith** [comment]
If I'm following you want to declaratively use that macro by checking if a flag is true and adding in data based on the condition?

I've ran into this sort of thing before and I ended up using additional filters for manipulating data within the nunjucks templates. 

So I expose something like lodash as nunjucks filters which can then be piped together. In this case might create the first part of the array a separate variable using `{% set %}` and then maybe do the conditional logic for the flag and do an array merge with the additional array items should the flag be true. 

Off the top of my head something like:

```
{% set radioItems = [
{
      value: "england",
      text: "England"
    },
    {
      value: "scotland",
      text: "Scotland"
    },
    {
      value: "wales",
      text: "Wales"
    },
    {
      value: "northern-ireland",
      text: "Northern Ireland"
    }
] | conditionalMerge(data.showFantasy, [
    {
      divider: "or"
    },
    {
      value: "narnia",
      text: "Narnia"
    }
] %}
```

**Comment by edwardhorsford** [comment]
@paulmsmith thanks for that!

For the context of this issue, I feel like the macro should be more defensive and not output anything where the input is blank. As a byproduct that also solves my issue ;)

Separately I'm somewhat sure I've seen a way of doing this in Nunjucks directly - but can't recall where / what it is.

**Comment by nickcolley** [comment]
I've had a go at this and it looks like a sensible suggestion to only output items when the required options have been supplied like Ed has suggested.

Will see what the rest of the team thinks, we'd want to do this for all the components I've mentioned above as well.

**Comment by paulmsmith** [comment]
I agree, the macro should be only outputting stuff when it has valid data. :)

**Comment by nickcolley** [comment]
Thanks Ed, We've added this to our backlog as part of the [improve macros in GOV.UK Frontend epic](https://github.com/alphagov/govuk-frontend/issues/1386).

If you wanted to contribute this sooner, feel free to raise a pull request.

**Comment by edwardhorsford** [comment]
A similar issue exists if you pass empty conditionals:
![empty-conditionals](https://user-images.githubusercontent.com/2204224/61715646-7186dd00-ad55-11e9-9e97-32f9d7ddad17.gif)

**Comment by edwardhorsford** [comment]
A similar thing would be good for tables - testing items inside `head` and `rows` if any are empty.

**Comment by 36degrees** [comment]
I believe this was addressed in #1512

**Comment by edwardhorsford** [comment]
@36degrees Should I raise my above issue with empty conditionals separately? or is that also solved with #1512

**Comment by 36degrees** [comment]
Ah, sorry @edwardhorsford, my bad. If you can raise a separate issue that'd be great – I don't believe it's solved by #1512 (though we do test for `{% if item.conditional %}`, so I'm not entirely sure what's going on there).
